### PR TITLE
Add Cache Manager support for Accept header

### PR DESCRIPTION
**Pull request by yadij**
This completes the support in cachemgr Action classes for
incremental addition of report formats other than text/plain
(historic cachemgr.cgi accepted syntax).

Reporters (class Action children) can now determine their
content-type support dynamically based on request
Accept header.

### CODE REVIEW

**Comment by yadij** [comment]
I have dropped all the YAML specific parts. This PR is now just making the Accept header available to report generators.

File: src/mgr/Action.h
Comment (rousskov) [inline src/mgr/Action.h:74]: Please undo this API change (and its effects). The undocumented (and currently unused) String parameter is not needed for the Action to compute the Content-Type value based on the command parameters (e.g., the Accept request header). The Mgr::Action::fillEntry() code in this PR demonstrates how the Accept request header info can be extracted when needed.

File: src/cache_manager.cc
Comment (rousskov) [inline src/cache_manager.cc:248]: > PR description: Reporters ... can now determine their content-type support dynamically and based on Accept header
inputs to something different from text/plain.

Something went wrong in that PR description edit -- the "to something different from text/plain" phrase is not attached to anything I can detect. I suggest to simplify: "Actions can now vary their responses based on the Accept request header."

Comment (rousskov) [inline src/cache_manager.cc:248]: This PR assumes that future Actions should react to Accept request header values, presumably producing output of two (or more) different Content-Types. Why is that a good idea? Should not all Actions simply respond using YAML instead, disregarding the Accept request header value (or validating it centrally, before the Action is even created and forwarded to workers)? If we have already agreed to proceed with this design, please feel free to point me to that agreement, of course.

**Review by rousskov — CHANGES_REQUESTED** [review]
As a workaround for a GitHub UI limitation, I am attaching two "global" change requests to the first (and innocent) line in the PR instead of listing them here, in this review comment, because change requests in such review comments are often ignored and their discussion is difficult to follow.

**Comment by rousskov** [comment]
I do not understand the current description[^1] of the `feature` label, but perhaps this PR does not need that mysterious label?

[^1]: `maintainer needs documentation update for merge`:  "For merge" of which PR? Where? It does not sound like we are talking about merging the labeled PR into its target branch... Please update the label name and description to clarify. If this label is about the need to update Squid release notes and such (after the PR is merged), then let's say that explicitly, preferably with a common reusable prefix! For example, `todo-release-notes` or `todo-update-release-docs`.

File: src/cache_manager.cc
Comment (yadij) [reply src/cache_manager.cc:248]: The ad-hoc format accepted by cachemgr.cgi is not fully compliant YAML. So to retain support for tools expecting that syntax we should only send YAML when it is indicated by the client Accept header.

IIRC, we agreed to the principal of moving to YAML, but still have disagreement on the specifics.

File: src/mgr/Action.h
Comment (yadij) [reply src/mgr/Action.h:74]: Hmm, it was needed for some reason I no longer recall. It is unused simply because the demo report conversion is gone.

Comment (yadij) [reply src/mgr/Action.h]: Done as of [8c9671e](https://github.com/squid-cache/squid/pull/70/commits/8c9671e8bf5843a22e2042a1e027feb67a9c5d96)

File: src/cache_manager.cc
Comment (rousskov) [reply src/cache_manager.cc:247]: > The ad-hoc format accepted by cachemgr.cgi is not fully compliant YAML.

Not sure why you are talking about cachemgr.cgi as if it specifies what output Squid does or should produce. That CGI script is just a hack that we fully control. It will accept whatever we make it accept. It does not and should not determine Squid output format(s). I do agree that Squid output is not YAML today, of course, but that fact has nothing to do with cachemgr.cgi.

> So to retain support for tools expecting that syntax we should only send YAML when it is indicated by the client Accept header.

IMO, we do not want to add code that works towards retaining support for tools requiring non-YAML syntax because we do not want to create and maintain code that generates responses in _two_ different formats, especially when one of those formats is an undefined/unstructured/ad-hoc/custom/informal format. We want to move towards YAML, not towards YAML+informal. We cannot get to YAML immediately, of course, but I see no value (and quite a bit of harm) in enabling this kind of Action code:

```C++
if (clientAcceptsYaml)
    generateYamlResponse();
else
    generateInformalResponse();
```

And if the above is not what we want, then I see no value in telling Actions what formats the client accepts.

> IIRC, we agreed to the principal of moving to YAML

Yes, that matches my recollection as well.

> but still have disagreement on the specifics.

That is probably true, but this change request disagreement appears to be about the very target/direction of that move rather than any move _specifics_. If we target just one format, then Actions do not need to know what formats the client accepts.

**Comment by yadij** [comment]
Cache manager is no longer planned to present anything other than YAML. Closing as unneeded

**Pull request closed without merging**
